src/libsystemd/sd-event/event-util.c

```diff
@@ -177,3 +177,85 @@ dual_timestamp* event_dual_timestamp_now(sd_event *e, dual_timestamp *ts) {
         assert_se(sd_event_now(e, CLOCK_MONOTONIC, &ts->monotonic) >= 0);
         return ts;
 }
+
+void event_source_unref_many(sd_event_source **array, size_t n) {
+        FOREACH_ARRAY(v, array, n)
+                sd_event_source_unref(*v);
+
+        free(array);
+}
+
+static int event_forward_signal_callback(sd_event_source *s, const struct signalfd_siginfo *ssi, void *userdata) {
+        sd_event_source *child = ASSERT_PTR(userdata);
+
+        assert(ssi);
+
+        siginfo_t si = {
+                .si_signo = ssi->ssi_signo,
+                /* This has to be SI_QUEUE as SI_USER and SI_KERNEL are not allowed to be used. */
```

Comment:
christian says SI_USER should probably work? did you test this?

Reply:
SI_USER doesn't actually work. You can't spoof it. But really it doesn't matter. If you want to give the other side a way to recognize that this was forwarded just pick a negative value like `INT_MIN`. The kernel doesn't validate this field much.